### PR TITLE
KafkaSpanCollector metrics support

### PR DESCRIPTION
**Pull request by kristofa**
Now that we have 2 `SpanCollector` implementations (Scribe and Kafka) I renamed `ScribeCollectorMetricsHandler` and moved it to `brave-core`module so it can also be used for the Kafka implementation.  This allows us to track number of accepted and dropped spans for all current and future SpanCollectors.

@pbetkier @adriancole

### CODE REVIEW

File: brave-core/src/main/java/com/github/kristofa/brave/SpanCollectorMetricsHandler.java [hunk 1]
```diff
@@ -1,14 +1,14 @@
-package com.github.kristofa.brave.scribe;
+package com.github.kristofa.brave;
 
 /**
- * Monitor {@link ScribeSpanCollector} by implementing reactions to these events, e.g. updating suitable metrics.
+ * Monitor SpanCollector by implementing reactions to these events, e.g. updating suitable metrics.
```
Comment (adriancole) [inline brave-core/src/main/java/com/github/kristofa/brave/SpanCollectorMetricsHandler.java]: supernit `{@linkplain SpanCollector}` to make this refactor-safe, or change to "span collector" to disconnect from type names.

File: brave-spancollector-kafka/README.md [hunk 2]
```diff
@@ -3,4 +3,14 @@
 SpanCollector that is used to submit spans to Kafka.
 
 Spans are sent to kafka as keyed messages: the key is the topic `zipkin` 
-and the value is a TBinaryProtocol encoded Span.
\ No newline at end of file
+and the value is a TBinaryProtocol encoded Span.
+
+## Monitoring ##
+
+Monitor `KafkaSpanCollector`'s by providing your implementation of `SpanCollectorMetricsHandler`. It will
```
Comment (adriancole) [inline brave-spancollector-kafka/README.md]: nice docs!

File: brave-spancollector-kafka/src/main/java/com/github/kristofa/brave/kafka/KafkaSpanCollector.java [hunk 3]
```diff
@@ -42,35 +43,43 @@ private static Properties defaultPropertiesWith(String bootstrapServers) {
...
     private final BlockingQueue<Span> queue;
+    private final SpanCollectorMetricsHandler metricsHandler;
 
     /**
      * Create a new instance with default configuration.
      *
      * @param bootstrapServers A list of host/port pairs to use for establishing the initial connection to the Kafka cluster.
      *                         Like: host1:port1,host2:port2,... Does not to be all the servers part of Kafka cluster.
+     * @param metricsHandler   Gets notified when spans are accepted or dropped. If you are not interested in these events you
+     *                         can use EmptySpanCollectorMetricsHandler
```
Comment (adriancole) [inline brave-spancollector-kafka/src/main/java/com/github/kristofa/brave/kafka/KafkaSpanCollector.java]: supernit: link or linkplain to make this refactor-safe

File: brave-spancollector-kafka/src/main/java/com/github/kristofa/brave/kafka/KafkaSpanCollector.java [hunk 4]
```diff
@@ -42,35 +43,43 @@ private static Properties defaultPropertiesWith(String bootstrapServers) {
...
      * KafkaSpanCollector.
      *
-     * @param kafkaProperties    Configuration for Kafka producer. Essential configuration properties are:
-     *                           bootstrap.servers, key.serializer, value.serializer. For a
-     *                           full list of config options, see http://kafka.apache.org/documentation.html#producerconfigs.
+     * @param kafkaProperties Configuration for Kafka producer. Essential configuration properties are:
+     *                        bootstrap.servers, key.serializer, value.serializer. For a
+     *                        full list of config options, see http://kafka.apache.org/documentation.html#producerconfigs.
+     * @param metricsHandler  Gets notified when spans are accepted or dropped. If you are not interested in these events you
+     *                        can use EmptySpanCollectorMetricsHandler
```
Comment (adriancole) [inline brave-spancollector-kafka/src/main/java/com/github/kristofa/brave/kafka/KafkaSpanCollector.java]: same

File: brave-spancollector-kafka/src/main/java/com/github/kristofa/brave/kafka/KafkaSpanCollector.java [hunk 5]
```diff
@@ -91,6 +100,7 @@ public void close() {
         }
         executorService.shutdown();
         producer.close();
+        metricsHandler.incrementDroppedSpans(queue.size());
```
Comment (adriancole) [inline brave-spancollector-kafka/src/main/java/com/github/kristofa/brave/kafka/KafkaSpanCollector.java]: edge case, but if we don't call `awaitTermination` or `isTerminated` after `shutdown`, would this value be accurate?

File: brave-spancollector-kafka/src/test/java/com/github/kristofa/brave/kafka/ITKafkaSpanCollector.java [hunk 6]
```diff
@@ -25,6 +21,24 @@
 public class ITKafkaSpanCollector {
 
     private final static String TOPIC = "zipkin";
+    private final EventsHandler metricsHandler = new EventsHandler();
+
+    private static class EventsHandler implements SpanCollectorMetricsHandler {
```
Comment (adriancole) [inline brave-spancollector-kafka/src/test/java/com/github/kristofa/brave/kafka/ITKafkaSpanCollector.java]: nice

**Comment by adriancole** [comment]
LGTM nice job

File: brave-core/src/test/java/com/github/kristofa/brave/DropwizardMetricsScribeCollectorMetricsHandlerExample.java [hunk 7]
```diff
@@ -1,8 +1,9 @@
-package com.github.kristofa.brave.scribe;
+package com.github.kristofa.brave;
 
 import com.codahale.metrics.MetricRegistry;
+import com.github.kristofa.brave.SpanCollectorMetricsHandler;
 
-class DropwizardMetricsScribeCollectorMetricsHandlerExample implements ScribeCollectorMetricsHandler {
+class DropwizardMetricsScribeCollectorMetricsHandlerExample implements SpanCollectorMetricsHandler {
```
Comment (pbetkier) [inline brave-core/src/test/java/com/github/kristofa/brave/DropwizardMetricsScribeCollectorMetricsHandlerExample.java]: I would say this class is no longer related to scribe, so we could rename it to `DropwizardMetricsSpanCollectorMetricsHandlerExample` and remove scribe keyword from the meter paths below. If you decide to do it, please note that this class name appears in readme files.

**Comment by pbetkier** [comment]
:+1:

File: brave-spancollector-kafka/src/main/java/com/github/kristofa/brave/kafka/KafkaSpanCollector.java (hunk 5 above)
Comment (kristofa) [inline brave-spancollector-kafka/src/main/java/com/github/kristofa/brave/kafka/KafkaSpanCollector.java]: I think what we have is correct because we wait for `SpanProcessingTask` to finish before calling `shutdown`. So I expect that after `shutdown` is called `isTerminated` should be true. 

In other words, once `SpanProcesssingTask` is finished the remaining items on the queue won't be processed anymore.

File: brave-core/src/test/java/com/github/kristofa/brave/DropwizardMetricsScribeCollectorMetricsHandlerExample.java (hunk 7 above)
Comment (kristofa) [inline brave-core/src/test/java/com/github/kristofa/brave/DropwizardMetricsScribeCollectorMetricsHandlerExample.java]: Oh yes. I overlooked this one. I'll fix it. Thanks!

File: brave-spancollector-kafka/src/main/java/com/github/kristofa/brave/kafka/KafkaSpanCollector.java (hunk 5 above)
Comment (adriancole) [inline brave-spancollector-kafka/src/main/java/com/github/kristofa/brave/kafka/KafkaSpanCollector.java]: Ah OK. Thx for the loopback
